### PR TITLE
[3.x] Compatibility with Laravel 9

### PR DESCRIPTION
**Pull request by erikgaal**
There were some more changes required before getting the package to work with Laravel 9. I've added testing on L9 to the workflow so we can be sure it's compatible before we merge.

Thank you @CodeSkills for the work in your PR. I've made sure to add you as a co-author for the commit regarding the `FilesystemTenancyBootstrapper`.

### CODE REVIEW

**Comment by stancl** [comment]
Thanks for the PRs! I'm happy to merge #800 or #802, whichever adds the L9 support fully

With regards to the workflow changes you made, I think it'll fail now since Laravel 9 requires PHP 8. And it's trying to use it with PHP 7.4.

It'd probably be best to merge #793 first. There are a few changes remaining there, so if you could try implementing them, we'll be able to see how this PR works.

I just responded to a few reviews there.

File: composer.json [hunk 1]
```diff
@@ -11,16 +11,16 @@
     ],
     "require": {
         "ext-json": "*",
-        "illuminate/support": "^6.0|^7.0|^8.0",
+        "illuminate/support": "^6.0|^7.0|^8.0|^9.0",
         "facade/ignition-contracts": "^1.0",
```
Comment (stancl) [inline composer.json:15]: Won't this have the same issue that @codeskills tried to address in his PR?

Comment (erikgaal) [reply composer.json:15]: Yes, but I think we should keep the contracts so the solutions work for older versions. I don't think there is a way to support solutions in libraries at the moment without including `ignition` in the project.

Comment (stancl) [reply composer.json:15]: If `facade/ignition-contracts` doesn't cause issues on L9, I agree that this makes sense. That seems to be the case, looking at the CI build.

If Spatie ends up making `spatie/ignition-contracts` **not** depend on entire ignition, we can try to see if there's a way to support both <L9 and L9

I'll keep this review open

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 2]
```diff
@@ -54,23 +54,22 @@ public function bootstrap(Tenant $tenant)
...
+
         foreach ($this->app['config']['tenancy.filesystem.disks'] as $disk) {
-            /** @var FilesystemAdapter $filesystemDisk */
-            $filesystemDisk = Storage::disk($disk);
-            $this->originalPaths['disks'][$disk] = $filesystemDisk->getAdapter()->getPathPrefix();
+            $root = $this->app['config']->get("filesystems.disks.{$disk}.root");
+            $this->originalPaths['disks'][$disk] = $root;
 
-            if ($root = str_replace(
+            if (!($root = str_replace(
```
Comment (stancl) [inline src/Bootstrappers/FilesystemTenancyBootstrapper.php]: @codeskills could you explain the changes a bit?

File: tests/BootstrapperTest.php [hunk 3]
```diff
@@ -187,27 +187,6 @@ public function filesystem_data_is_separated()
...
-            } else {
-                if ($base = $old_storage_facade_roots[$disk]) {
-                    $correct_path_prefix = $base . "/$suffix/";
-                } else {
-                    $correct_path_prefix = "$suffix/";
-                }
-            }
-
-            $this->assertSame($correct_path_prefix, $current_path_prefix);
-        }
```
Comment (stancl) [inline tests/BootstrapperTest.php]: Why was this removed?

File: src/Middleware/CheckTenantForMaintenanceMode.php [hunk 4]
```diff
@@ -29,7 +29,12 @@ public function handle($request, Closure $next)
                 return $next($request);
             }
 
-            throw new MaintenanceModeException($data['time'], $data['retry'], $data['message']);
+            throw new HttpException(
+                503,
+                'Service Unavailable',
+                null,
+                isset($data['retry']) ? ['Retry-After' => $data['retry']] : []
+            );
```
Comment (stancl) [inline src/Middleware/CheckTenantForMaintenanceMode.php:37]: Was `MaintenanceModeException` removed?

Comment (erikgaal) [reply src/Middleware/CheckTenantForMaintenanceMode.php:37]: It's deprecated since L8, and with Symfony 6 using it it throws a TypeError.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 5]
```diff
@@ -84,14 +83,9 @@ public function revert()
         $this->app['url']->setAssetRoot($this->app['config']['app.asset_url']);
 
         // Storage facade
+        Storage::forgetDisk($this->app['config']['tenancy.filesystem.disks']);
```
Comment (stancl) [inline src/Bootstrappers/FilesystemTenancyBootstrapper.php:87]: Is `forgetDisk()` used to solve the prefix issue without having to manipulate the existing disk instances?

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 2 above)
Comment (stancl) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: I'd like to make sure this uses the root overrides correctly, as described here https://tenancyforlaravel.com/docs/v2/filesystem-tenancy/

I think I might add some assertions to the `BootstrapperTest`, assuming the PR can be set up locally

Comment (erikgaal) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: I think this was an IDE suggestion to negate the condition and remove the original `if` part. Because the original `if` part of the conditional (the `setPathPrefix` call) was no longer needed.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 5 above)
Comment (erikgaal) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php:87]: Correct, it makes sure that the next time the disk is accessed, it is recreated from the new config.

File: tests/BootstrapperTest.php (hunk 3 above)
Comment (erikgaal) [reply tests/BootstrapperTest.php]: I feel like this part is unnecessarily testing the internals of the `FilesystemBootstrapper`. On line 189 the storage path is already tested by checking the result of the `storage_path` function. The test also broke down because we can't do `getPathPrefix` anymore. Testing the end-user results (`storage_path()` and `Storage::facade()`) makes more sense to me.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 2 above)
Comment (CodeSkills) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: It is just a simplification, it works the same way as before, since there is `$this->app['config']["filesystems.disks.{$disk}.root"] = $root;` at the end of the function. You dont have to do this `$filesystemDisk->getAdapter()->setPathPrefix($finalPrefix = $root);` because it changes the config itself.

Comment (stancl) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Pushed those test changes now

File: src/Middleware/CheckTenantForMaintenanceMode.php (hunk 4 above)
Comment (stancl) [reply src/Middleware/CheckTenantForMaintenanceMode.php:37]: And the code above is how Laravel implements maintenance mode in L8+?

I'd like to avoid making a breaking change here if people depend on the old exception in some way. But if the new implementation is similar to how Laravel does this, it's probably fine.

File: tests/BootstrapperTest.php (hunk 3 above)
Comment (stancl) [reply tests/BootstrapperTest.php]: I believe the point of this was to test that the actual adapter instances use the right paths.

`storage_path()` is irrelevant in that context, and `Storage::facade()` doesn't allow fetching the prefix. I pushed some tests now to assert that the paths used by the instances in `Storage` are correct

File: src/Middleware/CheckTenantForMaintenanceMode.php (hunk 4 above)
Comment (erikgaal) [reply src/Middleware/CheckTenantForMaintenanceMode.php:37]: Yeah, I copied this from Laravel's codebase :)

Comment (stancl) [reply src/Middleware/CheckTenantForMaintenanceMode.php:37]: Cool, we can go with this then

**Comment by erikgaal** [comment]
Looks good to me now :)

File: tests/BootstrapperTest.php (hunk 3 above)
Comment (erikgaal) [reply tests/BootstrapperTest.php]: Right, makes sense and should be fine this way now. I do believe a better abstraction of resolving the paths (and testing those) would result in a robuster implementation. Maybe something for v4.

**Comment by stancl** [comment]
Do you think we can finish 793 as I wrote here? https://github.com/archtechx/tenancy/pull/802#issuecomment-1039104506

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 2 above)
Comment (stancl) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: I'd rewrite the if to not include the negation and assignment in the condition. You can make them separate

**Comment by colinmackinlay** [comment]
> Do you think we can finish 793 as I wrote here? [#802 (comment)](https://github.com/archtechx/tenancy/pull/802#issuecomment-1039104506)

@stancl - would be great if this could be merged now if it is working. Even if you don't tag a release until #793 is done we could still pull it in via `dev:master` and run Laravel 9 upgrades :)

File: tests/BootstrapperTest.php [hunk 6]
```diff
@@ -184,30 +189,38 @@ public function filesystem_data_is_separated()
         $this->assertFalse(Storage::disk('public')->exists('foo'));
         $this->assertFalse(Storage::disk('public')->exists('abc'));
 
+        $expected_storage_path = $old_storage_path . '/tenant' . tenant('id'); // /tenant = suffix base
+
+        // Check that disk prefixes respect the root_override logic
+        $this->assertSame($expected_storage_path . '/app/', $this->getDiskPrefix('local'));
+        $this->assertSame($expected_storage_path . '/app/public/', $this->getDiskPrefix('public'));
+        $this->assertSame('tenant' . tenant('id') . '/', ltrim($this->getDiskPrefix('s3'), '/'));
```
Comment (stancl) [inline tests/BootstrapperTest.php:197]: I'm using `ltrim()` in the tests here, to make the strings identical, but per what @CodeSkills suggested in https://github.com/archtechx/tenancy/pull/800#issuecomment-1038433316 there might be a problem with the prefixing? So if this issue happens in the app and not just tests, the `/` prefix adjustment should be made in the bootstrapper rather than tests.

Comment (stancl) [reply tests/BootstrapperTest.php:197]: (For context, the tests don't use s3 at all. They just make some general config assertions.)

File: composer.json (hunk 1 above)
Comment (stancl) [reply composer.json:15]: Seems like the `ignition-contracts` dependency also makes the package incompatible with PHP 8.1? https://github.com/facade/ignition-contracts/blob/3c921a1cdba35b68a7f0ccffc6dffc1995b18267/composer.json#L20

**Comment by stancl** [comment]
@colinmackinlay I can merge the PR sooner and tag a version after all of the other things are resolved, but there are still some reviews here that should be addressed before merging.

File: composer.json (hunk 1 above)
Comment (arcdigital) [reply composer.json:15]: they have it set to `^8.0`, which will allow PHP 8.1 to work fine

Comment (stancl) [reply composer.json:15]: Ah, good catch. I think we'll stay with the old `ignition-contracts` package then, for now

**Comment by SupianIDz** [comment]
> @colinmackinlay I can merge the PR sooner and tag a version after all of the other things are resolved, but there are still some reviews here that should be addressed before merging.

is there any progress ? We have entered the tenancy stage of our program. I hope it's merged soon.

Thank you all for your hard work ❤️ .

**Comment by magarrent** [comment]
Hey! It's the package already compatible with L9? Or there are some pending issues? Maybe I can help

**Comment by stancl** [comment]
Would appreciate @CodeSkills' feedback on the review where I mentioned him, and a refactor of this https://github.com/archtechx/tenancy/pull/802/files/f5269b2b025e24243cba78e06435f25bfff9b3fa#diff-a7131a9582938c1b642f3d517a9064cbacb7b83d75fb52fe2c6e52dbca26a584

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 2 above)
Comment (CodeSkills) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: You can rewrite it to something like this, to make it more clear:
```php
$root = str_replace(
    '%storage_path%',
    storage_path(),
    $this->app['config']["tenancy.filesystem.root_override.{$disk}"] ?? ''
);
// Default root
if (!$root) {
    $root = "{$root}/{$suffix}";
}
```

**Comment by CodeSkills** [comment]
I have written there an update on how you could rewrite it. Hope it helps! (Someone should probably test it before since it has been changed from my PR, when somebody replaced the vars with {$root}, which I did not test.

File: tests/BootstrapperTest.php (hunk 6 above)
Comment (stancl) [reply tests/BootstrapperTest.php:197]: @CodeSkills could you comment on the `/` prefix changes? It'd help a lot here

**Comment by oxiginedev** [comment]
hey! @stancl , please what's the update on support for laravel 9. Any help needed?

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php (hunk 2 above)
Comment (erikgaal) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: I've refactored the proposed changes to prevent overwriting used variables. It took me a while to fully understand what the code was doing, but this should clear it up.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 7]
```diff
@@ -54,20 +54,20 @@ public function bootstrap(Tenant $tenant)
...
                 storage_path(),
-                $this->app['config']["tenancy.filesystem.root_override.{$disk}"] ?? ''
-            )) {
-                $filesystemDisk->getAdapter()->setPathPrefix($finalPrefix = $root);
-            } else {
-                $root = $this->app['config']["filesystems.disks.{$disk}.root"];
-                $filesystemDisk->getAdapter()->setPathPrefix($finalPrefix = $root . "/{$suffix}");
+                $this->app['config']["tenancy.filesystem.root_override.{$disk}"] ?? '',
+            );
+
```
Comment (vstyler96) [inline src/Bootstrappers/FilesystemTenancyBootstrapper.php:68]:
```suggestion

$finalPrefix = str_replace('%tenant%', $suffix, $finalPrefix);

```
I would like to suggest the following suggestion if this is accepted. let add a custom path with the tenant as argument.

Comment (stancl) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php:68]: If this is a new feature, like in the other PR, it's best to add it after merging this.

This PR should be just for the Laravel upgrade

**Comment by stancl** [comment]
I'll probably finish and merge this later today 👍🏻

**Comment by stancl** [comment]
Thanks for the $finalPrefix refactor (as well as your work on all this) @erikgaal! It looks great now.

The only thing I'm unsure about is the review re: S3, but I'll merge this into 3.x anyway. I'd appreciate if people who use S3 could test that branch and report if there are any issues. If there are not, I'll tag a release tomorrow. I'll also try to revisit the logic tomorrow (in code, not prod) to see if it's okay.

So to use the dev branch locally, you can just run:
```sh
composer require stancl/tenancy:3.x-dev
```

And then switch it back to `^3.0` once this is released.

**Comment by tkuhnert** [comment]
I just updated locally to L9 using 3.x-dev. Using flysystem-aws-s3-v3 in combination with a minio server. Everything works as expected so far. 👍

**Comment by stancl** [comment]
Thanks a lot for the testing!

**Comment by jtomek** [comment]
Hi, I also tested it on Vapor. All worked fine except the filesystem root url. I have described the issue and a hotfix here: https://github.com/archtechx/tenancy/issues/739#issuecomment-1067763548

**Comment by stancl** [comment]
@jtomek Can you confirm that this regression was introduced in the latest release (and therefore it all worked fine in 3.5.1)?

**Comment by jtomek** [comment]
@stancl, I will try my best to find a slot for it. This week is rather busy. I can share that I have been using and updating your package regularly for a long time and the issue at hand started with the Laravel 9 update.

**Comment by miketimeturner** [comment]
> @jtomek Can you confirm that this regression was introduced in the latest release (and therefore it all worked fine in 3.5.1)?

I can confirm this is the case. I have the same issue and therefor tenant directories now reside in directory "/" this is also having an impact retrieving the media and you are met with a permission denied, probably due to an incorrect url.

The issue no longer exist after reverting to 3.5.1

**Comment by stancl** [comment]
Thanks for the testing! I'll try to get it fixed this week.

**Comment by trippo** [comment]
Any news on this?
Are now production-ready v.3.5.2?

**Comment by stancl** [comment]
See the release notes

**Comment by trippo** [comment]
Thanks seem that v3.5.5 solves all bugs.

Thanks
